### ISSUE TITLE
Mirrored display on Unity 2019 & iOS

### ISSUE THREAD

**Issue by spazzout** [googlevr/gvr-unity-sdk#1031]
![IMG_DBC131DD29E3-1](https://user-images.githubusercontent.com/33397433/57441353-0e9fa280-724a-11e9-89f0-203bb79542f1.jpeg)

### SPECIFIC ISSUE ENCOUNTERED
The GVR camera transform rotation is inverted on Unity 2019.
Only happens on iOS builds. Both with single and multi pass.
Locking orientation doesn't make a difference either.

### HARDWARE/SOFTWARE VERSIONS
- Unity: 2019.1.0f2
- XCode: 10.2.1
- Google VR SDK for Unity: 1.18.5
- Tested on iOS 12.2:  iPhone 6s plus, iPhone 8, iPhone Xs, iPhone Xs plus
- Built on Mac mini: MacOS 10.14.4

### STEPS TO REPRODUCE THE ISSUE
1. Create a new unity project.
2. Set target platform to iOS.
3. Enable VR in build settings to import GVR package and add "cardboard" to devices.
4. Set orientation locked to landscape left
5. Build to iOS device and rotate it

### ADDITIONAL COMMENTS
Other fixes i have tried (but don't work) include:
- Inverting the camera rendering manually with GL, still the rotation applied to the VR camera is inverted.
- Inverting the transform rotation of the vr camera every frame has the same result.
- Setting different cardboard viewer presets by scanning their qr codes.

**Comment by spazzout** [comment]
Installed GVRSDK pod is 1.170

**Comment by spazzout** [comment]
So i tried downgrading the GVR iOS package from 1.18.5 down to 1.17 to match the podfile without improvement.

**Comment by futurnaingenieur** [comment]
Same issue here, build for Android:

![GoogleVRUnityIssue](https://user-images.githubusercontent.com/8960917/57569669-a216d600-73f8-11e9-9f19-cd1c90530573.png)

**Comment by niharsail** [comment]
Same thing happening with me.

**Comment by Rickleung** [comment]
Still no solution from unity?

**Comment by christougher** [comment]
Same Problems here for IOS.  It seems IOS 12.2 and or Unity 2019 have this problem.  IOS seems to do this even with Unity 2018

**Comment by christougher** [comment]
Seems to work fine under Graphics API OpenGLES3 but not Vulkan or Metal APIs...

**Comment by yazansorour** [comment]
downgrade unity to 2018.3.14f1 will works

**Comment by mariovski** [comment]
Go to player settings and remove Vulcan Api and stay with openGLES3.
It works fine now on 2019.1.2f1 and Android Huawai p20 lite and others.

**Comment by BinaryNate** [comment]
+1 for this issue. I'm unable to upgrade to Unity 2019.1 because enabling the Cardboard viewer in XR Settings causes the view to be upside-down on iOS. The issue is also [described in this Unity forum thread](https://answers.unity.com/questions/1627145/cardboard-google-vr-ios-results-in-upside-down-vie.html).

**Comment by spazzout** [comment]
**Temporary fix**
So we haven't heard back from Unity yet and needed to ship our game, my collegae came up with this script, attach it to your camera and it should render correctly on iOS.

[FlipScreen.txt](https://github.com/googlevr/gvr-unity-sdk/files/3284736/FlipScreen.txt)

**Comment by christougher** [comment]
Thanks so much for sharing this!  Not seeming to fix it for me...  Do we need to assign a material?  I added it to the Main Camera in the HelloVR example scene but no change.  Any other steps to follow? Thx again!

**Comment by spazzout** [comment]
Forgot to share this too, add a material with this shader to the script and it should work
[FlipShader.zip](https://github.com/googlevr/gvr-unity-sdk/files/3287139/FlipShader.zip)

**Comment by christougher** [comment]
Works great! Thx again.

**Comment by christougher** [comment]
Actually It appears that the left and right eye have about a 90 degree angle difference...

**Comment by eqrxjz** [comment]
I added to my project, works mostly fine, like @christougher mention, it has a disparity between cameras, maybe i can adjust that disparity in the script bottom section. i'll working on it, but by now it is the only workaround i found. Thanks! @spazzout

**Comment by juscelinotanaka** [comment]
I'm having the same problem. I am using the OpenGLES3 option for now. I can also add that when using Metal it keeps drawing a ScreenSpaceOverlay canvas over the inverted image while when using OpenGLES3 it doesn't show it (I know I shouldn't use it and/or expect it to work, I just made it for testing).

Before enabling VR:
![IMG_7642](https://user-images.githubusercontent.com/1385572/59720415-c54d5600-91ec-11e9-8bf3-38a677b31bd3.png)

After enabling VR - notice that it now has a slight alpha transparency/color multiplier:
![IMG_7643](https://user-images.githubusercontent.com/1385572/59720414-c4b4bf80-91ec-11e9-8b97-f31d71dc7321.png)

**Comment by paulgnz** [comment]
The Temporary fix above doesn't work with Metal. OpenGLES3 works however the performance is terrible.

Any suggestions?

**Comment by christougher** [comment]
Got an email from Unity QA...
"I have just verified the fixes and it should be available in the next released Unity version."

Not sure exactly which version(s) he's referring to... Hopefully 2018.4 as well as 2019.1...

**Comment by alcristog** [comment]
> Go to player settings and remove Vulcan Api and stay with openGLES3.
> It works fine now on 2019.1.2f1 and Android Huawai p20 lite and others.
> https://computergraphics.stackexchange.com/questions/8185/what-are-metal-and-vulcan

This worked for me!

**Comment by iamtheonly** [comment]
I'm using ARKit and Google Cardboard in my project. ARKit requires Metal, so I can't remove it and use GLES3 :(

I can confirm this does work in Unity 2019.3a, but then my project does not upload to App Store because of some Unity problems - "disallowed Frameworks", so I have to stick to 2019.1

Is this such a big issue that the fix can't be back ported to 2019.1 ? :/

**Comment by xinaesthete** [comment]
@iamtheonly have you tried the workaround scripts posted above?

**Comment by iamtheonly** [comment]
@xinaesthete yeah I did, and as someone posted it produces a huge disparity between the eyes. I haven't looked into the code too, maybe I can fix it myself.

I'm using cardboard for a 360 panorama viewer so I figured that simply flipping the texture upside down in the unity's panoramic shader will do (and it somewhat does - it displays correct!), but then when I build the app to the phone, the camera movement on X axis is inverted, so when I tilt the device down, the camera points upwards and vice versa.

**Comment by alcristog** [comment]
I got rid of any complication using the 2017 version for all my VR projects. This SDK seems not to work properly on 2019 versions...

**Comment by xinaesthete** [comment]
@iamtheonly ah ok. I haven't actually tried the workarounds I mentioned yet.  I do see similar things with the transforms going the wrong, flipping a texture upside-down definitely not going to fix it (actually the app I tried it with is sufficiently abstract that you can only tell it's wrong because the movement is wrong).

FWIW it seems fine with 2018 @alcristog.

**Comment by pauluhn** [comment]
2019.1.10f1 doesn't seem to work with the flip hack. the left/right aren't aligned correctly.

**Comment by christougher** [comment]
I can confirm the fix is included in 2019.2. Building fine on iOS.

**Comment by angamesahmed** [comment]
if any one still facing this, i have done simple solution
- put the VR camera under empty object and set that object z rotation to 180f
- make a simple script that read the local euler y rotation from camera.
- reverse the rotation it by multiply to -2f and set to y rotation of the camera parent empty object, and make that script in LateUpdate()
Wish it help others, as i tried most 2019 versions and the issue still exist for iOS
` using UnityEngine;
 [RequireComponent(typeof(Camera))]
 //[ExecuteInEditMode]
 public class MirrorFlipCamera : MonoBehaviour {
     new Camera camera;

    Transform origin;
    public bool invertCull;
     void Awake () {
         camera = GetComponent<Camera>();
        origin = transform.parent;
     }


    Vector3 correctedRotation = Vector3.zero;
    void LateUpdate()
    {
        //correctedRotation = origin.localEulerAngles;
        correctedRotation.y = camera.transform.localEulerAngles.y * -2f;
        origin.localEulerAngles = correctedRotation;
    }
 }`

**Comment by zxcvb7653** [comment]
> I can confirm the fix is included in 2019.2. Building fine on iOS.

Ditto
